test/integration_test/pxrepo_test.go

```diff
@@ -0,0 +1,129 @@
+			}
+			cluster.Spec.DeleteStrategy = &corev1.StorageClusterDeleteStrategy{
+				Type: corev1.UninstallAndWipeStorageClusterStrategyType,
+			}
+			return cluster
+		},
+		TestFunc: func(tc *types.TestCase) func(*testing.T) {
+			return func(t *testing.T) {
+				testSpec := tc.TestSpec(t)
+				cluster, ok := testSpec.(*corev1.StorageCluster)
+				require.True(t, ok)
+
+				// Populate default values to empty fields first
+				portworx.SetPortworxDefaults(cluster)
+				// Record pre-deploy timestamp
+				installTime := time.Now()
+				// Deploy cluster
+				_, err := ci_utils.CreateStorageCluster(cluster)
+				require.NoError(t, err)
+
+				// Validate cluster deployment
+				// -- WARNING: please make sure kernel sources/headers are not installed on the nodes, or the test will fail
+				// -- e.g. `cd /usr/src; dpkg --purge linux-headers-generic linux-generic *`
+				logrus.Infof("Validate StorageCluster %s has failed events", cluster.Name)
+				err = test.ValidateStorageClusterFailedEvents(cluster, ci_utils.DefaultValidateDeployTimeout, ci_utils.DefaultValidateDeployRetryInterval, "reason=FileSystemDependency", installTime, "")
+				require.NoError(t, err)
+
+				ci_utils.UninstallAndValidateStorageCluster(cluster, t)
+			}
+		},
+		ShouldSkip: func(tc *types.TestCase) bool {
+			// need px-operator v1.7, and px v2.10 or higher
+			return ci_utils.PxOperatorVersion.LessThan(ci_utils.PxOperatorVer1_7) ||
+				ci_utils.GetPxVersionFromSpecGenURL(ci_utils.PxSpecGenURL).LessThan(pxVer2_10)
+		},
+	},
+
+	{
+		TestName:        "InstallAirGappedWithPXRepo",
+		TestrailCaseIDs: []string{"TODO", "TODO2"},
+		TestSpec: func(t *testing.T) interface{} {
+			cluster := &corev1.StorageCluster{}
+			cluster.Name = "px-repo-included"
+			err := ci_utils.ConstructStorageCluster(cluster, ci_utils.PxSpecGenURL, ci_utils.PxSpecImages)
+			require.NoError(t, err)
+			testEnv := []string{
+				"PX_IMAGE=docker.io/zoxpx/enterprise:2.9.1-fixd",
+				// $TEST_SKIP_EXTRACT_PX_MODULE prevents unpacking of internal KO-archive
+				"TEST_SKIP_EXTRACT_PX_MODULE=true",
+				// set up pre-exec to clean up various px.ko module locations
+				"PRE-EXEC=rmmod -f px ; rm -fr /opt/pwx/bin/px.ko /opt/pwx/bin/px.ko.xz /var/lib/osd/pxfs /opt/pwx/oci/rootfs/home/px-fuse/px.ko",
+			}
+			for _, s := range testEnv {
+				sp := strings.Split(s, "=")
+				cluster.Spec.Env = append(cluster.Spec.Env, k8sv1.EnvVar{Name: sp[0], Value: sp[1]})
+			}
+			// Adding `-air-gapped` param stops download of kernel-headers and/or KO-modules from mirrors
+			cluster.Annotations = map[string]string{
+				"portworx.io/misc-args": "-air-gapped",
+			}
+			// Initiate PX-Repo container
+			cluster.Spec.PxRepo = &corev1.PxRepoSpec{
+				Enabled: true,
+				Image:   "docker.io/portworx/px-repo:latest",
+			}
+			cluster.Spec.DeleteStrategy = &corev1.StorageClusterDeleteStrategy{
+				Type: corev1.UninstallAndWipeStorageClusterStrategyType,
+			}
+			return cluster
+		},
+		TestFunc: BasicInstall,
```

Comment:
In jenkins the test case would run on non airgapped env, could you add validation that modules are installed through px-repo not public?

Reply:
This is OK -- the tests are using internal args to "emulate" the `air-gapped` env:
* `env TEST_SKIP_EXTRACT_PX_MODULE=true` -- this skips unpacking the internal KO-archive embedded in `px-enterprise` container
* `-air-gapped` flag added -- this prevents any downloads from mirrors.portworx.com or yum.portworx.com  (e.g. remote download and install of KO-modules and kernel-headers)

Reply:
how do we know if px-repo is used? my theory for a happy path validation is running with a px version that does not work, with a wrong px-repo image px wont run, with a correct px-repo image px runs correctly.

Reply:
The first (negative) test must fail with `Failed to load PX filesystem dependencies for kernel XXX` error -- the second test should pass, after we added the px-repo into the install.